### PR TITLE
Lberg/add chop

### PR DESCRIPTION
**Pull request by lucabergamini**
Add function to chop dataset (employed to generate private test set and GT csv).

Also update notebook to use this function during eval and plot only specific frames

### CODE REVIEW

File: examples/agent_motion_prediction/agent_motion_prediction.ipynb [hunk 1]
```diff
@@ -191,7 +193,23 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# Evaluation"
+    "# Evaluation\n",
+    "\n",
+    "Evaluation follows a slightly different protocol from training. When working with time series, we must be absolutely sure to avoid leaking the future in the data.\n",
```
Comment (oliver-scheel) [inline examples/agent_motion_prediction/agent_motion_prediction.ipynb:198]: than training

File: examples/agent_motion_prediction/agent_motion_prediction.ipynb [hunk 2]
```diff
@@ -191,7 +193,23 @@
...
+    "\n",
+    "If we followed the same protocol of training, one could just read ahead in the `.zarr` and forge a perfect solution at run-time, even for a private test set.\n",
+    "\n",
+    "As such, **the private test set for the competition has been \"chopped\" using the `chop_dataset` function**.\n",
+    "\n",
+    "The result is that each scene has been reduced to only 100 frames, and only valid agents in the 100th frame will be used to compute the metrics. Because following frames in the scene have been chopped off, we can't just look ahead to get the future of those agents.\n",
```
Comment (oliver-scheel) [inline examples/agent_motion_prediction/agent_motion_prediction.ipynb:204]: Is this really how it's done, only valid agents in the 100th frame - not valid agents in any of the first 100 frames?

File: examples/agent_motion_prediction/agent_motion_prediction.ipynb [hunk 3]
```diff
@@ -191,7 +193,23 @@
...
+    "\n",
+    "In this example, we simulate this pipeline by running `chop_dataset` on the validation set. The function stores:\n",
+    "- a new chopped `.zarr` dataset where each scene has only the first 100 frames;\n",
```
Comment (oliver-scheel) [inline examples/agent_motion_prediction/agent_motion_prediction.ipynb:207]: , in which each scene only contains the first 100 frames

File: examples/agent_motion_prediction/agent_motion_prediction.ipynb [hunk 4]
```diff
@@ -191,7 +193,23 @@
...
+    "- a numpy mask array where only valid agents in the 100th frame are True;\n",
+    "- a ground-truth file with the future coordinates of those agents;\n",
+    "\n",
+    "During the competition, only the `.zarr` and the mask will be provided for the private test set evaluation.\n",
+    "Your solution is expected to generate a csv file which will be compared with the ground truth one on a separate server"
```
Comment (oliver-scheel) [inline examples/agent_motion_prediction/agent_motion_prediction.ipynb:212]: file, which ... compared to the ..

File: l5kit/l5kit/dataset/chop_dataset.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+MIN_FUTURE_STEPS = 10
+FUTURE_STEPS = 50
+
+
+def create_chopped_dataset(
+    zarr_path: str, th_agent_prob: float, num_frames_to_copy: int, num_frames_gt: int, min_frame_future: int
+) -> str:
+    """
+    Create a chopped version of the zarr that can be used for evaluation. This function is used to generate tests
```
Comment (oliver-scheel) [inline l5kit/l5kit/dataset/chop_dataset.py:20]: test sets

File: l5kit/l5kit/dataset/select_agents.py [hunk 6]
```diff
@@ -231,7 +231,7 @@ def select_agents(
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
     parser.add_argument("--input_folders", nargs="+", type=str, required=True, help="zarr path")
-    parser.add_argument("--th_agent_prob", type=float, default=0.5, help="perception threshold on agents of interest")
+    parser.add_argument("--th_agent_prob", type=float, required=True, help="perception threshold on agents")
```
Comment (oliver-scheel) [inline l5kit/l5kit/dataset/select_agents.py:234]: Any reason for this?

File: l5kit/l5kit/dataset/chop_dataset.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        frame_original = zarr_dt.frames[scene["frame_index_interval"][0] + num_frames_to_copy - 1]
+        slice_agents_original = slice(*frame_original["agent_index_interval"])
+        frame_chopped = zarr_chopped.frames[zarr_chopped.scenes[idx]["frame_index_interval"][-1] - 1]
+        slice_agents_chopped = slice(*frame_chopped["agent_index_interval"])
+
+        mask = agents_mask_origin[slice_agents_original][:, 1] >= min_frame_future
+        agents_mask_orig_bool[slice_agents_original] = mask.copy()
+        agents_mask_chop_bool[slice_agents_chopped] = mask.copy()
+
+    # store the mask and the GT csv for frames on interest
```
Comment (oliver-scheel) [inline l5kit/l5kit/dataset/chop_dataset.py:80]: of

File: l5kit/l5kit/dataset/chop_dataset.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+from l5kit.data import ChunkedDataset
+from l5kit.data.zarr_utils import zarr_scenes_chunk
+from l5kit.dataset.select_agents import TH_DISTANCE_AV, TH_EXTENT_RATIO, TH_YAW_DEGREE, select_agents
+from l5kit.evaluation import export_zarr_to_csv
+
+MIN_FUTURE_STEPS = 10
+FUTURE_STEPS = 50
+
+
+def create_chopped_dataset(
```
Comment (oliver-scheel) [inline l5kit/l5kit/dataset/chop_dataset.py:16]: I'm a bit confused about the naming here: if I understand it correctly, the chopped zarr (dataset) is actually created by zarr_scenes_chunk(), whereas here we 'just' compute the mask and export the GT csv?

File: l5kit/l5kit/dataset/chop_dataset.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    Create a chopped version of the zarr that can be used for evaluation. This function is used to generate tests
+    set that have no GT in the data for selected agents.
+
+    Store:
+     - a dataset where each scene has been chopped at `num_frames_to_copy` frames;
+     - a mask for agents for those final frames based on the original mask and a threshold on the future_frames;
+     - the GT csv for those agents
+
+    Args:
+        zarr_path (str): zarr path as a string
```
Comment (oliver-scheel) [inline l5kit/l5kit/dataset/chop_dataset.py:29]: Minor, but I feel these comments are a bit hastily written: 'threshold over agents prob ...', 'min number of frames available in the future for an agent to be taken'

File: examples/agent_motion_prediction/agent_motion_prediction.ipynb (hunk 2 above)
Comment (lucabergamini) [reply examples/agent_motion_prediction/agent_motion_prediction.ipynb:204]: nope, otherwise you leak the future when you include the following frames

File: l5kit/l5kit/dataset/select_agents.py (hunk 6 above)
Comment (lucabergamini) [reply l5kit/l5kit/dataset/select_agents.py:234]: Because `th_agent_prob` is used for the folder name, it makes little sense to leave it with a default value. When you run this script that param is the one you want to move

File: l5kit/l5kit/dataset/chop_dataset.py (hunk 8 above)
Comment (lucabergamini) [reply l5kit/l5kit/dataset/chop_dataset.py:16]: no, we also store the chopped dataset here. But I agree that we should change chunk to chop everywhere for clarity.
This function is a high level abstraction that puts together a couple of other utilities.
It is included only to be transparent about how we build the chopped dataset....I'm not really sure this is the right place for it in fact, what do you think @oliver-scheel ?

File: examples/visualisation/visualise_data.ipynb [hunk 10]
```diff
@@ -65,7 +65,7 @@
    "outputs": [],
    "source": [
     "# set env variable for data\n",
-    "os.environ[\"L5KIT_DATA_FOLDER\"] = \"PATH_TO_YOUR_DATA\"\n",
+    "os.environ[\"L5KIT_DATA_FOLDER\"] = \"/tmp/l5kit_data\"\n",
```
Comment (oliver-scheel) [inline examples/visualisation/visualise_data.ipynb:68]: Is this change on purpose?

File: l5kit/l5kit/dataset/chop_dataset.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+from l5kit.evaluation import export_zarr_to_csv
+
+MIN_FUTURE_STEPS = 10
+
+
+def create_chopped_dataset(
+    zarr_path: str, th_agent_prob: float, num_frames_to_copy: int, num_frames_gt: int, min_frame_future: int
+) -> str:
+    """
+    Create a chopped version of the zarr that can be used as a test set. This function is used to generate tests
```
Comment (oliver-scheel) [inline l5kit/l5kit/dataset/chop_dataset.py:19]: still: tests set -> test sets

File: l5kit/l5kit/dataset/chop_dataset.py (hunk 8 above)
Comment (oliver-scheel) [reply l5kit/l5kit/dataset/chop_dataset.py:15]: That makes sense. Minor, but in line 58 there is a comment # create chopped dataset and then zarr_scenes_chop is called -> this was my thinking, the comments with the name of this function points to zarr_scenes_chop - but I am fine with the current version.
Regarding where to put it, yeah, this does not actually need to be in L5Kit right? No one uses it, but it is just for showing how we do it. It is probably still nice to keep it as it belongs here - where: I also do not have a better idea. One could think of a folder 'competition', but if this is the only script in there, it does not fit better than here imo.

File: examples/agent_motion_prediction/agent_motion_prediction.ipynb [hunk 12]
```diff
@@ -47,7 +49,7 @@
    "outputs": [],
    "source": [
     "# set env variable for data\n",
-    "os.environ[\"L5KIT_DATA_FOLDER\"] = \"PATH_TO_YOUR_DATA\"\n",
+    "os.environ[\"L5KIT_DATA_FOLDER\"] = \"/tmp/l5kit_data\"\n",
```
Comment (stefaniespeichert) [inline examples/agent_motion_prediction/agent_motion_prediction.ipynb:52]: Maybe change back to the generic one?

File: l5kit/l5kit/dataset/chop_dataset.py (hunk 8 above)
Comment (lucabergamini) [reply l5kit/l5kit/dataset/chop_dataset.py:15]: we use it the prediction notebook to simulate what will happen during the competition, but we could have also just shipped the validation set already chopped with the mask and the gt....

File: examples/visualisation/visualise_data.ipynb (hunk 10 above)
Comment (lucabergamini) [reply examples/visualisation/visualise_data.ipynb:68]: Obviously no xD It's the 100th time I fix this

File: l5kit/l5kit/dataset/chop_dataset.py (hunk 11 above)
Comment (lucabergamini) [reply l5kit/l5kit/dataset/chop_dataset.py:19]: Everything is singular now in that sentence. Do you want everything plural?

Comment (oliver-scheel) [reply l5kit/l5kit/dataset/chop_dataset.py:19]: For me the second sentence still is 'This function is used to generate tests set that ...'

Comment (lucabergamini) [reply l5kit/l5kit/dataset/chop_dataset.py:19]: hahah sorry I thought you were referring to the first one :)
